src/libs/actions/Policy.js

```diff
@@ -1458,6 +1461,375 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
     return lodashUnion([category], policyRecentlyUsedCategories);
 }
 
+/**
+ * This flow is used for bottom up flow converting IOU report to an expense report. When user takes this action,
+ * we create a Collect type workspace when the person taking the action becomes an owner and an admin, while we
+ * add a new member to the workspace as an employee and convert the IOU report passed as a param into an expense report.
+ *
+ * @param {Object} iouReport
+ * @returns {String} policyID of the workspace we have created
+ */
+function createWorkspaceFromIOUPayment(iouReport) {
+    // This flow only works for IOU reports
+    if (!ReportUtils.isIOUReport(iouReport)) {
+        return;
+    }
+
+    // Generate new variables for the policy
+    const policyID = generatePolicyID();
+    const workspaceName = generateDefaultWorkspaceName(sessionEmail);
+    const employeeAccountID = iouReport.ownerAccountID;
+    const employeeEmail = iouReport.ownerEmail;
+    const {customUnits, customUnitID, customUnitRateID} = buildOptimisticCustomUnits();
+    const oldPersonalPolicyID = iouReport.policyID;
+    const iouReportID = iouReport.reportID;
+
+    const {
+        announceChatReportID,
+        announceChatData,
+        announceReportActionData,
+        announceCreatedReportActionID,
+        adminsChatReportID,
+        adminsChatData,
+        adminsReportActionData,
+        adminsCreatedReportActionID,
+        expenseChatReportID: workspaceChatReportID,
+        expenseChatData: workspaceChatData,
+        expenseReportActionData: workspaceChatReportActionData,
+        expenseCreatedReportActionID: workspaceChatCreatedReportActionID,
+    } = ReportUtils.buildOptimisticWorkspaceChats(policyID, workspaceName);
+
+    // Create the workspace chat for the employee whose IOU is being paid
+    const employeeWorkspaceChat = createPolicyExpenseChats(policyID, {[employeeEmail]: employeeAccountID});
+    const newWorkspace = {
+        id: policyID,
+        type: CONST.POLICY.TYPE.TEAM, // We are creating a collect policy in this case
+        name: workspaceName,
+        role: CONST.POLICY.ROLE.ADMIN,
+        owner: sessionEmail,
+        isPolicyExpenseChatEnabled: true,
+        outputCurrency: CONST.CURRENCY.USD, // Setting the currency to USD as we can only add the VBBA for this policy currency right now
+        pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD,
+        customUnits,
+    };
+
+    const optimisticData = [
+        {
+            onyxMethod: Onyx.METHOD.SET,
```

Comment:
I think it should, although I did take this from the CreateWorkspace so not sure if there were any problems with using merge, I will try

Reply:
I actually had a GH at one point to go and fix a lot of that `CreateWorkspace` stuff to use merge, but I wasn't ever able to prioritize it. It is a bad pattern that shouldn't be followed due to the inconsistencies of working with both merge/set and many times `set()` is used as a workaround hack for things like wanting the operation to be sync instead of async, and not taking the time to make what is relying on the data being in Onyx to handle the async operation.

Reply:
I think we should look at other solutions. The purpose of `SET` isn't to fix navigation issues.